In [89]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/contraceptive/contraceptive prevalence any method married women age 15-49.csv
/kaggle/input/contraceptive/contraceptive any method all women 15-49.csv
/kaggle/input/contraceptive/contraceptive prevalence any method unmarried women age 15-49 .csv


**Analyzing Global Contraceptive Prevalence: Trends, Disparities, and Socioeconomic Impacts**

**Research Objectives**

To analyze trends in contraceptive prevalence across different groups (all women, married women, and sexually active unmarried women).

To compare contraceptive prevalence across regions and income levels.

To assess how socioeconomic factors (e.g., education, income, healthcare access) influence contraceptive use.

To evaluate disparities in contraceptive access between married and unmarried women.


**Research Questions**

How has contraceptive prevalence changed globally over the years for different population groups?

What are the regional disparities in contraceptive use among married and unmarried women?

How does contraceptive prevalence correlate with fertility rate, education, and GDP per capita?

What factors contribute to lower contraceptive prevalence among certain groups?


**Data downloaded from World Bank Group**

Women data is for all (regardless of married or unmarried status)

married data is for married ladies and unmarried is for sexually active unmarried women

In [90]:
#Load  data
import pandas as pd

women = pd.read_csv("/kaggle/input/contraceptive/contraceptive any method all women 15-49.csv")
married = pd.read_csv("/kaggle/input/contraceptive/contraceptive prevalence any method married women age 15-49.csv")
unmarried = pd.read_csv("/kaggle/input/contraceptive/contraceptive prevalence any method unmarried women age 15-49 .csv")

In [91]:
women.head()

Series Name        Series Code  \
0  Contraceptive prevalence, any method (% of all...  SP.DYN.CONU.AL.ZS   
1  Contraceptive prevalence, any method (% of all...  SP.DYN.CONU.AL.ZS   
2  Contraceptive prevalence, any method (% of all...  SP.DYN.CONU.AL.ZS   
3  Contraceptive prevalence, any method (% of all...  SP.DYN.CONU.AL.ZS   
4  Contraceptive prevalence, any method (% of all...  SP.DYN.CONU.AL.ZS   

                  Country Name Country Code 1960 [YR1960] 1961 [YR1961]  \
0                  Afghanistan          AFG            ..            ..   
1  Africa Eastern and Southern          AFE            ..            ..   
2   Africa Western and Central          AFW            ..            ..   
3                      Albania          ALB            ..            ..   
4                      Algeria          DZA            ..            ..   

  1962 [YR1962] 1963 [YR1963] 1964 [YR1964] 1965 [YR1965]  ... 2014 [YR2014]  \
0            ..            ..            ..            ..  ...            ..   
1            ..            ..            ..            ..  ...            ..   
2            ..            ..            ..            ..  ...            ..   
3            ..            ..            ..            ..  ...            ..   
4            ..            ..            ..            ..  ...            ..   

  2015 [YR2015] 2016 [YR2016] 2017 [YR2017]     2018 [YR2018] 2019 [YR2019]  \
0            ..            ..            ..                ..            ..   
1            ..            ..            ..  34.5437689482571            ..   
2            ..            ..            ..  17.3750237165794            ..   
3            ..            ..            ..              33.2            ..   
4            ..            ..            ..                ..            ..   

  2020 [YR2020] 2021 [YR2021] 2022 [YR2022] 2023 [YR2023]  
0            ..            ..            ..            ..  
1            ..            ..            ..            ..  
2            ..            ..            ..            ..  
3            ..            ..            ..            ..  
4            ..            ..            ..            ..  

[5 rows x 68 columns]

In [92]:
married.head()

Series Name     Series Code  \
0  Contraceptive prevalence, any method (% of mar...  SP.DYN.CONU.ZS   
1  Contraceptive prevalence, any method (% of mar...  SP.DYN.CONU.ZS   
2  Contraceptive prevalence, any method (% of mar...  SP.DYN.CONU.ZS   
3  Contraceptive prevalence, any method (% of mar...  SP.DYN.CONU.ZS   
4  Contraceptive prevalence, any method (% of mar...  SP.DYN.CONU.ZS   

                  Country Name Country Code 1960 [YR1960] 1961 [YR1961]  \
0                  Afghanistan          AFG            ..            ..   
1  Africa Eastern and Southern          AFE            ..            ..   
2   Africa Western and Central          AFW            ..            ..   
3                      Albania          ALB            ..            ..   
4                      Algeria          DZA            ..            ..   

  1962 [YR1962] 1963 [YR1963] 1964 [YR1964] 1965 [YR1965]  ... 2014 [YR2014]  \
0            ..            ..            ..            ..  ...            ..   
1            ..            ..            ..            ..  ...            ..   
2            ..            ..            ..            ..  ...            ..   
3            ..            ..            ..            ..  ...            ..   
4            ..            ..            ..            ..  ...            ..   

  2015 [YR2015] 2016 [YR2016] 2017 [YR2017] 2018 [YR2018]     2019 [YR2019]  \
0          22.5          22.5            ..            ..                ..   
1            ..            ..            ..            ..  42.3600499131128   
2            ..            ..            ..            ..  19.1597537149594   
3            ..            ..            ..            46                ..   
4            ..            ..            ..            ..              53.6   

  2020 [YR2020] 2021 [YR2021] 2022 [YR2022] 2023 [YR2023]  
0            ..            ..            ..            ..  
1            ..            ..            ..            ..  
2            ..            ..            ..            ..  
3            ..            ..            ..            ..  
4            ..            ..            ..            ..  

[5 rows x 68 columns]

In [93]:
unmarried.head()

Series Name        Series Code  \
0  Contraceptive prevalence, any method (% of sex...  SP.DYN.CONU.SA.ZS   
1  Contraceptive prevalence, any method (% of sex...  SP.DYN.CONU.SA.ZS   
2  Contraceptive prevalence, any method (% of sex...  SP.DYN.CONU.SA.ZS   
3  Contraceptive prevalence, any method (% of sex...  SP.DYN.CONU.SA.ZS   
4  Contraceptive prevalence, any method (% of sex...  SP.DYN.CONU.SA.ZS   

                  Country Name Country Code 1960 [YR1960] 1961 [YR1961]  \
0                  Afghanistan          AFG            ..            ..   
1  Africa Eastern and Southern          AFE            ..            ..   
2   Africa Western and Central          AFW            ..            ..   
3                      Albania          ALB            ..            ..   
4                      Algeria          DZA            ..            ..   

  1962 [YR1962] 1963 [YR1963] 1964 [YR1964] 1965 [YR1965]  ... 2014 [YR2014]  \
0            ..            ..            ..            ..  ...            ..   
1            ..            ..            ..            ..  ...            ..   
2            ..            ..            ..            ..  ...            ..   
3            ..            ..            ..            ..  ...            ..   
4            ..            ..            ..            ..  ...            ..   

  2015 [YR2015] 2016 [YR2016] 2017 [YR2017]    2018 [YR2018] 2019 [YR2019]  \
0            ..            ..            ..               ..            ..   
1            ..            ..            ..               ..            ..   
2            ..            ..            ..  37.078655528028            ..   
3            ..            ..            ..             60.5            ..   
4            ..            ..            ..               ..            ..   

  2020 [YR2020] 2021 [YR2021] 2022 [YR2022] 2023 [YR2023]  
0            ..            ..            ..            ..  
1            ..            ..            ..            ..  
2            ..            ..            ..            ..  
3            ..            ..            ..            ..  
4            ..            ..            ..            ..  

[5 rows x 68 columns]

In [94]:
print(women.columns.tolist())  


['Series Name', 'Series Code', 'Country Name', 'Country Code', '1960 [YR1960]', '1961 [YR1961]', '1962 [YR1962]', '1963 [YR1963]', '1964 [YR1964]', '1965 [YR1965]', '1966 [YR1966]', '1967 [YR1967]', '1968 [YR1968]', '1969 [YR1969]', '1970 [YR1970]', '1971 [YR1971]', '1972 [YR1972]', '1973 [YR1973]', '1974 [YR1974]', '1975 [YR1975]', '1976 [YR1976]', '1977 [YR1977]', '1978 [YR1978]', '1979 [YR1979]', '1980 [YR1980]', '1981 [YR1981]', '1982 [YR1982]', '1983 [YR1983]', '1984 [YR1984]', '1985 [YR1985]', '1986 [YR1986]', '1987 [YR1987]', '1988 [YR1988]', '1989 [YR1989]', '1990 [YR1990]', '1991 [YR1991]', '1992 [YR1992]', '1993 [YR1993]', '1994 [YR1994]', '1995 [YR1995]', '1996 [YR1996]', '1997 [YR1997]', '1998 [YR1998]', '1999 [YR1999]', '2000 [YR2000]', '2001 [YR2001]', '2002 [YR2002]', '2003 [YR2003]', '2004 [YR2004]', '2005 [YR2005]', '2006 [YR2006]', '2007 [YR2007]', '2008 [YR2008]', '2009 [YR2009]', '2010 [YR2010]', '2011 [YR2011]', '2012 [YR2012]', '2013 [YR2013]', '2014 [YR2014]', '2

In [95]:
#Strip Any Extra Spaces from Column Names
women.columns = women.columns.str.strip()
married.columns = women.columns.str.strip()
unmarried.columns = unmarried.columns.str.strip()


In [96]:
# Keep only relevant columns
columns_to_drop = ["Series Name", "Series Code", "Country Code"]  

women = women.drop(columns=[col for col in columns_to_drop if col in women.columns])
married = married.drop(columns=[col for col in columns_to_drop if col in married.columns])
unmarried = unmarried.drop(columns=[col for col in columns_to_drop if col in unmarried.columns])


In [97]:
#check if the drop was successful
print(women.columns.tolist())  
print(married.columns.tolist())  
print(unmarried.columns.tolist())  


['Country Name', '1960 [YR1960]', '1961 [YR1961]', '1962 [YR1962]', '1963 [YR1963]', '1964 [YR1964]', '1965 [YR1965]', '1966 [YR1966]', '1967 [YR1967]', '1968 [YR1968]', '1969 [YR1969]', '1970 [YR1970]', '1971 [YR1971]', '1972 [YR1972]', '1973 [YR1973]', '1974 [YR1974]', '1975 [YR1975]', '1976 [YR1976]', '1977 [YR1977]', '1978 [YR1978]', '1979 [YR1979]', '1980 [YR1980]', '1981 [YR1981]', '1982 [YR1982]', '1983 [YR1983]', '1984 [YR1984]', '1985 [YR1985]', '1986 [YR1986]', '1987 [YR1987]', '1988 [YR1988]', '1989 [YR1989]', '1990 [YR1990]', '1991 [YR1991]', '1992 [YR1992]', '1993 [YR1993]', '1994 [YR1994]', '1995 [YR1995]', '1996 [YR1996]', '1997 [YR1997]', '1998 [YR1998]', '1999 [YR1999]', '2000 [YR2000]', '2001 [YR2001]', '2002 [YR2002]', '2003 [YR2003]', '2004 [YR2004]', '2005 [YR2005]', '2006 [YR2006]', '2007 [YR2007]', '2008 [YR2008]', '2009 [YR2009]', '2010 [YR2010]', '2011 [YR2011]', '2012 [YR2012]', '2013 [YR2013]', '2014 [YR2014]', '2015 [YR2015]', '2016 [YR2016]', '2017 [YR2017]

****

Now that the columns are cleaned, let's reshape your datasets from a wide format (where each year is a separate column) to a long format (where each row represents a country, year, and contraceptive prevalence).



In [98]:
# Remove any extra spaces from column names
women.columns = women.columns.str.strip()
married.columns = married.columns.str.strip()
unmarried.columns = unmarried.columns.str.strip()


In [99]:
# Rename columns to standardize country column name if necessary
women.rename(columns={'Country Name': 'Country'}, inplace=True)
married.rename(columns={'Country Name': 'Country'}, inplace=True)
unmarried.rename(columns={'Country Name': 'Country'}, inplace=True)


In [100]:
women.head()

Country 1960 [YR1960] 1961 [YR1961] 1962 [YR1962]  \
0                  Afghanistan            ..            ..            ..   
1  Africa Eastern and Southern            ..            ..            ..   
2   Africa Western and Central            ..            ..            ..   
3                      Albania            ..            ..            ..   
4                      Algeria            ..            ..            ..   

  1963 [YR1963] 1964 [YR1964] 1965 [YR1965] 1966 [YR1966] 1967 [YR1967]  \
0            ..            ..            ..            ..            ..   
1            ..            ..            ..            ..            ..   
2            ..            ..            ..            ..            ..   
3            ..            ..            ..            ..            ..   
4            ..            ..            ..            ..            ..   

  1968 [YR1968]  ... 2014 [YR2014] 2015 [YR2015] 2016 [YR2016] 2017 [YR2017]  \
0            ..  ...            ..            ..            ..            ..   
1            ..  ...            ..            ..            ..            ..   
2            ..  ...            ..            ..            ..            ..   
3            ..  ...            ..            ..            ..            ..   
4            ..  ...            ..            ..            ..            ..   

      2018 [YR2018] 2019 [YR2019] 2020 [YR2020] 2021 [YR2021] 2022 [YR2022]  \
0                ..            ..            ..            ..            ..   
1  34.5437689482571            ..            ..            ..            ..   
2  17.3750237165794            ..            ..            ..            ..   
3              33.2            ..            ..            ..            ..   
4                ..            ..            ..            ..            ..   

  2023 [YR2023]  
0            ..  
1            ..  
2            ..  
3            ..  
4            ..  

[5 rows x 65 columns]

In [101]:
# Function to clean year columns by removing 'YR' from the column names
def clean_year_columns(df):
    # Ensure columns that should not be changed are excluded (e.g., 'Country Name')
    year_columns = df.columns[df.columns.str.contains(r'\[YR\d{4}\]')]  # Find year columns
    
    # Remove 'YR' and the brackets from the year column names
    df.columns = df.columns.str.replace(r'\[YR(\d{4})\]', r'\1', regex=True)
    
    return df

# Clean the year columns for each dataset
women= clean_year_columns(women)
married = clean_year_columns(married)
unmarried= clean_year_columns(unmarried)

# Verify the cleaned columns
print(women.head())

                       Country 1960 1960 1961 1961 1962 1962 1963 1963  \
0                  Afghanistan        ..        ..        ..        ..   
1  Africa Eastern and Southern        ..        ..        ..        ..   
2   Africa Western and Central        ..        ..        ..        ..   
3                      Albania        ..        ..        ..        ..   
4                      Algeria        ..        ..        ..        ..   

  1964 1964 1965 1965 1966 1966 1967 1967 1968 1968  ... 2014 2014 2015 2015  \
0        ..        ..        ..        ..        ..  ...        ..        ..   
1        ..        ..        ..        ..        ..  ...        ..        ..   
2        ..        ..        ..        ..        ..  ...        ..        ..   
3        ..        ..        ..        ..        ..  ...        ..        ..   
4        ..        ..        ..        ..        ..  ...        ..        ..   

  2016 2016 2017 2017         2018 2018 2019 2019 2020 2020 2021 2021  \
0

In [102]:
# Check the type and shape of 'women' after loading or processing
print(type(women))  # Check if it's still a DataFrame or numpy array
print(women)        # Print the content to see if it's a DataFrame or single value


<class 'pandas.core.frame.DataFrame'>
                         Country 1960 1960 1961 1961 1962 1962 1963 1963  \
0                    Afghanistan        ..        ..        ..        ..   
1    Africa Eastern and Southern        ..        ..        ..        ..   
2     Africa Western and Central        ..        ..        ..        ..   
3                        Albania        ..        ..        ..        ..   
4                        Algeria        ..        ..        ..        ..   
..                           ...       ...       ...       ...       ...   
266                          NaN       NaN       NaN       NaN       NaN   
267                          NaN       NaN       NaN       NaN       NaN   
268                          NaN       NaN       NaN       NaN       NaN   
269                          NaN       NaN       NaN       NaN       NaN   
270                          NaN       NaN       NaN       NaN       NaN   

    1964 1964 1965 1965 1966 1966 1967 1967 1968 

In [103]:
print(married.head())

                       Country 1960 1960 1961 1961 1962 1962 1963 1963  \
0                  Afghanistan        ..        ..        ..        ..   
1  Africa Eastern and Southern        ..        ..        ..        ..   
2   Africa Western and Central        ..        ..        ..        ..   
3                      Albania        ..        ..        ..        ..   
4                      Algeria        ..        ..        ..        ..   

  1964 1964 1965 1965 1966 1966 1967 1967 1968 1968  ... 2014 2014 2015 2015  \
0        ..        ..        ..        ..        ..  ...        ..      22.5   
1        ..        ..        ..        ..        ..  ...        ..        ..   
2        ..        ..        ..        ..        ..  ...        ..        ..   
3        ..        ..        ..        ..        ..  ...        ..        ..   
4        ..        ..        ..        ..        ..  ...        ..        ..   

  2016 2016 2017 2017 2018 2018         2019 2019 2020 2020 2021 2021  \
0

**Handle Missing Values**

**Check Missing Data Patterns**

** Check Missing Values as a Percentage**

In [104]:
# Drop duplicate column names
women.columns = pd.Series(women.columns).str.split().str[0]  # Take only the first part




In [105]:
# Check missing values percentage per column
missing_percentage = (women.isnull().sum() / len(women)) * 100
print(missing_percentage)


Country    1.845018
1960       1.845018
1961       1.845018
1962       1.845018
1963       1.845018
             ...   
2019       1.845018
2020       1.845018
2021       1.845018
2022       1.845018
2023       1.845018
Length: 65, dtype: float64


In [106]:
# Check missing values percentage per column
missing_percentage1 = (married.isnull().sum() / len(women)) * 100
print(missing_percentage1)


Country      1.845018
1960 1960    1.845018
1961 1961    1.845018
1962 1962    1.845018
1963 1963    1.845018
               ...   
2019 2019    2.583026
2020 2020    2.583026
2021 2021    2.583026
2022 2022    2.583026
2023 2023    2.583026
Length: 65, dtype: float64


In [107]:
# Check missing values percentage per column
missing_percentage3 = (unmarried.isnull().sum() / len(women)) * 100
print(missing_percentage)


Country    1.845018
1960       1.845018
1961       1.845018
1962       1.845018
1963       1.845018
             ...   
2019       1.845018
2020       1.845018
2021       1.845018
2022       1.845018
2023       1.845018
Length: 65, dtype: float64


This output means that 1.85% of values are missing in every column, including "Country", "1960", "1961", etc., up to "2023". Dataset appears to have a multivariate time-series format, where each row represents a country, and each column represents a year.



**Visualizing Missing Data **

In [108]:
# Correlation of missing values
print(married.isnull().corr())


            Country  1960 1960  1961 1961  1962 1962  1963 1963  1964 1964  \
Country    1.000000   1.000000   1.000000   1.000000   1.000000   1.000000   
1960 1960  1.000000   1.000000   1.000000   1.000000   1.000000   1.000000   
1961 1961  1.000000   1.000000   1.000000   1.000000   1.000000   1.000000   
1962 1962  1.000000   1.000000   1.000000   1.000000   1.000000   1.000000   
1963 1963  1.000000   1.000000   1.000000   1.000000   1.000000   1.000000   
...             ...        ...        ...        ...        ...        ...   
2019 2019  0.841995   0.841995   0.841995   0.841995   0.841995   0.841995   
2020 2020  0.841995   0.841995   0.841995   0.841995   0.841995   0.841995   
2021 2021  0.841995   0.841995   0.841995   0.841995   0.841995   0.841995   
2022 2022  0.841995   0.841995   0.841995   0.841995   0.841995   0.841995   
2023 2023  0.841995   0.841995   0.841995   0.841995   0.841995   0.841995   

           1965 1965  1966 1966  1967 1967  1968 1968  ...  201

What This Correlation Means

1️⃣ Perfect Correlation (1.0) Across Most Columns

Dataset contains duplicate or highly similar columns.

If two columns are identical, their correlation is 1.0.

If they are scaled versions of each other, correlation remains 1.0.

2️⃣ Strange Column Labels (1960 1960, 2019 2019)

Dataset might have column names duplicated accidentally (e.g., 1960 appearing twice).


3️⃣ A Large Portion of the Dataset Has 0.841995 Correlation

This suggests some form of constant proportionality between those variables.

Possible reasons:

Columns represent time series trends that move together.

Data is structured or formula-driven rather than naturally varying.

In [109]:
#Fix: Rename columns correctly to ensure uniqueness:
married.columns = [col.split()[0] if col.isdigit() else col for col in married.columns]



In [110]:
#Check for Duplicate Columns
duplicates = married.T.duplicated()
print(married.columns[duplicates])


Index(['2023 2023'], dtype='object')


In [111]:
#remove duplicate
married = married.loc[:, ~married.T.duplicated()]


In [112]:
#ensure data is numeric
print(married.dtypes)
married= married.apply(pd.to_numeric, errors='coerce')  # Convert non-numeric values


Country      object
1960 1960    object
1961 1961    object
1962 1962    object
1963 1963    object
              ...  
2018 2018    object
2019 2019    object
2020 2020    object
2021 2021    object
2022 2022    object
Length: 64, dtype: object


In [113]:
#Recalculate Correlation After Cleaning
corr_matrix = married.corr()
print(corr_matrix)


           Country  1960 1960  1961 1961  1962 1962  1963 1963  1964 1964  \
Country        NaN        NaN        NaN        NaN        NaN        NaN   
1960 1960      NaN        NaN        NaN        NaN        NaN        NaN   
1961 1961      NaN        NaN        NaN        NaN        NaN        NaN   
1962 1962      NaN        NaN        NaN        NaN        NaN        NaN   
1963 1963      NaN        NaN        NaN        NaN        1.0        NaN   
...            ...        ...        ...        ...        ...        ...   
2018 2018      NaN        NaN        NaN        NaN        NaN        NaN   
2019 2019      NaN        NaN        NaN        NaN        NaN        NaN   
2020 2020      NaN        NaN        NaN        NaN        NaN        NaN   
2021 2021      NaN        NaN        NaN        NaN        NaN        NaN   
2022 2022      NaN        NaN        NaN        NaN        NaN        NaN   

           1965 1965  1966 1966  1967 1967  1968 1968  ...  2013 2013  \
Co

/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()


That means the correlation matrix is now more realistic instead of showing 1.0 everywhere. However, the high correlations (0.98, 0.96, etc.) suggest that  dataset still has strongly correlated values across years.

Possible Reasons & Next Steps

✅ 1. Check for Duplicate Columns
Even though we fixed column names, there might still be duplicate data.

In [114]:
married = married.loc[:, ~married.columns.duplicated()]


Check for Highly Similar Data Across Years

If values don’t change much across years, they will naturally have high correlation. Check:

In [115]:
married.describe().T


/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()


count       mean        std   min       25%    50%        75%  \
Country      0.0        NaN        NaN   NaN       NaN    NaN        NaN   
1960 1960    0.0        NaN        NaN   NaN       NaN    NaN        NaN   
1961 1961    1.0  42.000000        NaN  42.0  42.00000  42.00  42.000000   
1962 1962    0.0        NaN        NaN   NaN       NaN    NaN        NaN   
1963 1963    2.0  33.350000  15.909903  22.1  27.72500  33.35  38.975000   
...          ...        ...        ...   ...       ...    ...        ...   
2018 2018   41.0  45.521951  20.981272  10.9  27.20000  46.00  62.100000   
2019 2019   52.0  49.652300  20.616303   6.9  32.63024  53.10  63.392283   
2020 2020   20.0  43.495000  20.611378  11.5  25.35000  41.45  64.350000   
2021 2021    7.0  42.842857  23.995893  11.0  24.75000  49.70  58.450000   
2022 2022    0.0        NaN        NaN   NaN       NaN    NaN        NaN   

                 max  
Country          NaN  
1960 1960        NaN  
1961 1961  42.000000  
1962 1962        NaN  
1963 1963  44.600000  
...              ...  
2018 2018  82.300000  
2019 2019  81.122224  
2020 2020  77.400000  
2021 2021  72.800000  
2022 2022        NaN  

[64 rows x 8 columns]

In [116]:
married.corr()

/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.10/dist-packages/pan

Country  1960 1960  1961 1961  1962 1962  1963 1963  1964 1964  \
Country        NaN        NaN        NaN        NaN        NaN        NaN   
1960 1960      NaN        NaN        NaN        NaN        NaN        NaN   
1961 1961      NaN        NaN        NaN        NaN        NaN        NaN   
1962 1962      NaN        NaN        NaN        NaN        NaN        NaN   
1963 1963      NaN        NaN        NaN        NaN        1.0        NaN   
...            ...        ...        ...        ...        ...        ...   
2018 2018      NaN        NaN        NaN        NaN        NaN        NaN   
2019 2019      NaN        NaN        NaN        NaN        NaN        NaN   
2020 2020      NaN        NaN        NaN        NaN        NaN        NaN   
2021 2021      NaN        NaN        NaN        NaN        NaN        NaN   
2022 2022      NaN        NaN        NaN        NaN        NaN        NaN   

           1965 1965  1966 1966  1967 1967  1968 1968  ...  2013 2013  \
Country          NaN        NaN        NaN        NaN  ...        NaN   
1960 1960        NaN        NaN        NaN        NaN  ...        NaN   
1961 1961        NaN        NaN        NaN        NaN  ...        NaN   
1962 1962        NaN        NaN        NaN        NaN  ...        NaN   
1963 1963        NaN        NaN        NaN        NaN  ...        NaN   
...              ...        ...        ...        ...  ...        ...   
2018 2018       -1.0       -1.0        NaN        NaN  ...   0.986791   
2019 2019        NaN        NaN        NaN        NaN  ...   0.982002   
2020 2020        NaN        NaN        NaN        NaN  ...   0.991036   
2021 2021        NaN        NaN        NaN        NaN  ...   1.000000   
2022 2022        NaN        NaN        NaN        NaN  ...        NaN   

           2014 2014  2015 2015  2016 2016  2017 2017  2018 2018  2019 2019  \
Country          NaN        NaN        NaN        NaN        NaN        NaN   
1960 1960        NaN        NaN        NaN        NaN        NaN        NaN   
1961 1961        NaN        NaN        NaN        NaN        NaN        NaN   
1962 1962        NaN        NaN        NaN        NaN        NaN        NaN   
1963 1963  -1.000000        NaN        NaN        NaN        NaN        NaN   
...              ...        ...        ...        ...        ...        ...   
2018 2018   0.960005   0.983580   0.990439   0.985496   1.000000   0.998470   
2019 2019   0.976222   0.988763   0.988577   0.931901   0.998470   1.000000   
2020 2020   0.903632   0.962948   0.979743   0.997264   0.993249   0.991158   
2021 2021   0.965716   0.923942   0.928912   0.914716   0.981834   1.000000   
2022 2022        NaN        NaN        NaN        NaN        NaN        NaN   

           2020 2020  2021 2021  2022 2022  
Country          NaN        NaN        NaN  
1960 1960        NaN        NaN        NaN  
1961 1961        NaN        NaN        NaN  
1962 1962        NaN        NaN        NaN  
1963 1963        NaN        NaN        NaN  
...              ...        ...        ...  
2018 2018   0.993249   0.981834        NaN  
2019 2019   0.991158   1.000000        NaN  
2020 2020   1.000000   0.993441        NaN  
2021 2021   0.993441   1.000000        NaN  
2022 2022        NaN        NaN        NaN  

[64 rows x 64 columns]

In [117]:
# Correlation of missing values
print(unmarried.isnull().corr())


           Country  1960 1960  1961 1961  1962 1962  1963 1963  1964 1964  \
Country        1.0        1.0        1.0        1.0        1.0        1.0   
1960 1960      1.0        1.0        1.0        1.0        1.0        1.0   
1961 1961      1.0        1.0        1.0        1.0        1.0        1.0   
1962 1962      1.0        1.0        1.0        1.0        1.0        1.0   
1963 1963      1.0        1.0        1.0        1.0        1.0        1.0   
...            ...        ...        ...        ...        ...        ...   
2019 2019      1.0        1.0        1.0        1.0        1.0        1.0   
2020 2020      1.0        1.0        1.0        1.0        1.0        1.0   
2021 2021      1.0        1.0        1.0        1.0        1.0        1.0   
2022 2022      1.0        1.0        1.0        1.0        1.0        1.0   
2023 2023      1.0        1.0        1.0        1.0        1.0        1.0   

           1965 1965  1966 1966  1967 1967  1968 1968  ...  2014 2014  \
Co

A correlation of 1.0 between all columns suggests that every column is perfectly correlated with every other column.

This is typically not expected in real-world data, unless:

Dataset has duplicate columns.

The data is structured incorrectly (e.g., columns contain the same values).

There is an issue with preprocessing (e.g., all missing values replaced with the same number).

There is a data leakage issue, meaning all variables carry the same information.


In [118]:
#Rename columns correctly to ensure uniqueness:
unmarried.columns = unmarried.columns.str.split().str[0]  # Keep only the first part of column names



In [119]:
#check for duplicate columns
duplicates = unmarried.T.duplicated().sum()
print(f"Number of duplicate columns: {duplicates}")


Number of duplicate columns: 28


In [120]:
#remove duplicate
unmarried = unmarried.loc[:, ~unmarried.T.duplicated()]


Data Might Be Categorical Instead of Numeric

If your dataset has strings or categorical data instead of numeric values, correlation can be misleading.

✅ Fix: Check data types:

In [ ]:
print(unmarried.dtypes)


In [ ]:
#If any column is not numeric, convert it
unmarried = unmarried.apply(pd.to_numeric, errors='coerce')


3️⃣ All Columns Have the Same Scaling or Constant Ratio

If every column follows the same transformation (e.g., all are scaled versions of each other), correlation might still be 1.0.

✅ Fix: Check for constant ratios:

In [122]:
unmarried_ratio = unmarried.div(unmarried.iloc[:, 0], axis=0)
print(unmarried_ratio.nunique())


TypeError: unsupported operand type(s) for /: 'str' and 'str'

If all columns have only one unique ratio, they are likely scaled versions of each other.


Dataset Might Have Been Accidentally Transposed

If rows and columns got swapped, it could lead to an all-1 correlation matrix.

✅ Fix: Check dimensions:
python
Copy
Edit


In [123]:
print(unmarried.shape)


(271, 37)


In [124]:
corr_matrix = unmarried.corr()
print(corr_matrix)


ValueError: could not convert string to float: 'Afghanistan'

In [125]:
#correlation for women
corr_matrix = women.corr()
print(corr_matrix)


ValueError: could not convert string to float: 'Afghanistan'

Little MCAR test

In [126]:
import numpy as np
import pandas as pd
import scipy.stats as stats

women = pd.read_csv("/kaggle/input/contraceptive/contraceptive any method all women 15-49.csv")
married = pd.read_csv("/kaggle/input/contraceptive/contraceptive prevalence any method married women age 15-49.csv")
unmarried = pd.read_csv("/kaggle/input/contraceptive/contraceptive prevalence any method unmarried women age 15-49 .csv")

def little_mcar_test(df):
    """Perform Little's MCAR test manually."""
    
    # Select only numerical columns (MCAR test requires numerical data)
    df_numeric = df.select_dtypes(include=[np.number])
    
    # Drop rows where all values are missing
    df_numeric = df_numeric.dropna(how="all")
    
    # Mean imputation (replacing missing values with column means)
    df_filled = df_numeric.fillna(df_numeric.mean())
    
    # Compute chi-square test
    s, p_value = stats.chisquare(df_numeric.isnull().sum(), df_filled.isnull().sum())
    
    # Print results
    print(f"MCAR Test p-value: {p_value:.4f}")
    if p_value > 0.05:
        print("Fail to reject MCAR -> Missing data is MCAR (Completely at Random).")
    else:
        print("Reject MCAR -> Missing data is MAR or MNAR (Not Completely at Random).")

# Run the function for each dataset
little_mcar_test(women)
little_mcar_test(married)
little_mcar_test(unmarried)


MCAR Test p-value: nan
Reject MCAR -> Missing data is MAR or MNAR (Not Completely at Random).
MCAR Test p-value: nan
Reject MCAR -> Missing data is MAR or MNAR (Not Completely at Random).
MCAR Test p-value: nan
Reject MCAR -> Missing data is MAR or MNAR (Not Completely at Random).


In [127]:
print(women.isnull().sum())


Series Name      3
Series Code      5
Country Name     5
Country Code     5
1960 [YR1960]    5
                ..
2019 [YR2019]    5
2020 [YR2020]    5
2021 [YR2021]    5
2022 [YR2022]    5
2023 [YR2023]    5
Length: 68, dtype: int64


Check If Missingness Depends on the Country (MAR Test)

If missing values in year are related to specific countries, the data is likely MAR.



Reshape Data (Convert Wide to Long Format)

To analyze MAR, we need to reshape the data from wide format to long format, where each row represents a Country-Year pair.

📌 Convert Wide to Long Format

In [128]:
import pandas as pd
women= pd.read_csv("/kaggle/input/contraceptive/contraceptive any method all women 15-49.csv")
married= pd.read_csv('/kaggle/input/contraceptive/contraceptive prevalence any method married women age 15-49.csv')
unmarried = pd.read_csv("/kaggle/input/contraceptive/contraceptive prevalence any method unmarried women age 15-49 .csv")
def reshape_data_to_long(df, id_col='Country Name'):
    return df.melt(id_vars=[id_col], var_name="Year", value_name="Value")


# Example usage:
women_long = reshape_data_to_long(women)
# Ensure "Value" remains as it is (without forcing float conversion)
women_long["Value"] = pd.to_numeric(women_long["Value"], errors="coerce")  # Convert valid numbers, keep NaNs


married_long = reshape_data_to_long(married)
# Ensure "Value" remains as it is (without forcing float conversion)
married_long["Value"] = pd.to_numeric(married_long["Value"], errors="coerce")  # Convert valid numbers, keep NaNs

unmarried_long = reshape_data_to_long(unmarried)
# Ensure "Value" remains as it is (without forcing float conversion)
unmarried_long["Value"] = pd.to_numeric(unmarried_long["Value"], errors="coerce")  # Convert valid numbers, keep NaNs


Why?

This makes it easier to check missingness patterns in relation to Country and Year.


 Check If Missingness is Related to Country (MAR Test)
 
If missing values in years (e.g., 1990, 2005) are more frequent for certain countries, this suggests MAR.

📌 Check Missingness by Country

In [129]:
import numpy as np
print(women_long.replace("", np.nan).isnull().sum())


Country Name      335
Year                0
Value           17417
dtype: int64


Check If Missingness is Related to Country (MAR Test)

If missing values in years (e.g., 1990, 2005) are more frequent for certain countries, this suggests MAR.

📌 Check Missingness by Country

In [130]:
def check_missing_by_country(df, country_col='Country Name', year_col='Year', value_col='Value'):
    missing_counts = df[df[value_col].isnull()].groupby(country_col).size()
    total_counts = df.groupby(country_col).size()
    
    missing_percentage = (missing_counts / total_counts) * 100
    missing_percentage = missing_percentage.dropna().sort_values(ascending=False)
    
    print(f"Percentage of Missing {value_col} by {country_col}:\n")
    print(missing_percentage)

# Example usage:
check_missing_by_country(women_long)
check_missing_by_country(married_long)
check_missing_by_country(unmarried_long)


Percentage of Missing Value by Country Name:

Country Name
Afghanistan         100.000000
OECD members        100.000000
Macao SAR, China    100.000000
Finland             100.000000
Fiji                100.000000
                       ...    
Kenya                80.597015
Philippines          79.104478
United Kingdom       79.104478
Ghana                76.119403
Peru                 70.149254
Length: 266, dtype: float64
Percentage of Missing Value by Country Name:

Country Name
Latin America & Caribbean    100.000000
Cayman Islands               100.000000
Isle of Man                  100.000000
Indicator Name               100.000000
Iceland                      100.000000
                                ...    
Korea, Rep.                   70.149254
Peru                          67.164179
Viet Nam                      65.671642
Japan                         62.686567
Indonesia                     53.731343
Length: 268, dtype: float64
Percentage of Missing Value by Country Name:


Interpretation of Results:

Country Variable:

From the output, the percentage of missing values for different countries varies. Some countries have 100% missing values for the data points, while others have varying degrees of missing data, with Kenya and Philippines showing around 80-90% missing data.

100% Missing Data: Countries like Afghanistan, OECD members, Macao SAR, and Finland show 100% missing data, which means no data was reported for those countries in the dataset for the variable in question.

Missing Data for Other Countries: Countries like Kenya, Philippines, United Kingdom, and Peru have missing data percentages ranging from 70% to 80%.

MAR Test Interpretation for Country:

The countries with 100% missing data (e.g., Afghanistan, Finland, etc.) suggest that missing data might not be missing at random because the missingness could be attributed to certain reasons (e.g., lack of data reporting, inconsistent collection sources).

For the countries with lower percentages of missing data (Kenya, Philippines, etc.), the missingness might be linked to specific years (e.g., years of missing reports), and this could indicate Missing at Random (MAR).

The MAR hypothesis suggests that the probability of missingness depends on observed data, and it’s possible that these countries have missing data for specific years rather than having a random pattern.

Conclusion for Country:

Since there are countries with 100% missing data and others with varying degrees of missingness, it suggests that the data for these countries is likely not Missing Completely at Random (MCAR), but it might follow a Missing at Random (MAR) pattern depending on observed data (such as reporting inconsistencies across countries).


Step 3: Check If Missingness is Related to Year

If missing values increase for certain years (e.g., more missing data in older years like 1961-1980), this could suggest MAR.

📌 Check Missingness by Year

Check If Missingness is Related to Year

If missing values increase for certain years (e.g., more missing data in older years like 1961-1980), this could suggest MAR.



In [131]:
def check_missing_by_year(df, year_col='Year', value_col='Value'):
    missing_counts = df[df[value_col].isnull()].groupby(year_col).size()
    total_counts = df.groupby(year_col).size()
    
    missing_percentage = (missing_counts / total_counts) * 100
    missing_percentage = missing_percentage.dropna().sort_values(ascending=False)
    
    print(f"Percentage of Missing {value_col} by {year_col}:\n")
    print(missing_percentage)

# Example usage:
check_missing_by_year(women_long)
check_missing_by_year(married_long)
check_missing_by_year(unmarried_long)


Percentage of Missing Value by Year:

Year
1960 [YR1960]    100.000000
1973 [YR1973]    100.000000
Series Code      100.000000
Country Code     100.000000
2023 [YR2023]    100.000000
                    ...    
2000 [YR2000]     87.822878
2006 [YR2006]     86.715867
2010 [YR2010]     86.346863
2014 [YR2014]     85.977860
2018 [YR2018]     81.918819
Length: 67, dtype: float64
Percentage of Missing Value by Year:

Year
1960 [YR1960]    100.000000
2022 [YR2022]    100.000000
Series Code      100.000000
Country Code     100.000000
2023 [YR2023]    100.000000
                    ...    
2014 [YR2014]     82.783883
2019 [YR2019]     80.952381
2006 [YR2006]     78.754579
2010 [YR2010]     75.091575
2000 [YR2000]     64.102564
Length: 67, dtype: float64
Percentage of Missing Value by Year:

Year
1960 [YR1960]    100.000000
1977 [YR1977]    100.000000
Series Code      100.000000
Country Code     100.000000
2023 [YR2023]    100.000000
                    ...    
1998 [YR1998]     96.678967
2016 

Year Variable:

The percentage of missing values for the years also shows significant variability across different years:

100% Missing Data for Certain Years: Some years (1960, 2023) have 100% missing data, which means no values were recorded for those years in the dataset for the variable in question.

Missing Data for Other Years: Other years such as 2000, 2006, 2010, 2014, etc., have missing data percentages ranging from 80% to 90%.

MAR Test Interpretation for Year:
For years with 100% missing data (e.g., 1960, 2023), missing data is systematic and might be due to data availability issues for those specific years. This could suggest a Not Missing at Random (NMAR) scenario for those years since there’s no information at all for certain time periods.

The years with missing data percentages ranging from 80% to 90% (e.g., 2000, 2006, 2010) may be Missing at Random (MAR), where the missingness is likely dependent on other observed variables (like the country or the availability of reports for those years).

Conclusion for Year:

Similar to the country variable, the year data is likely not Missing Completely at Random (MCAR) because of the presence of 100% missing years.
However, for other years with a lower percentage of missing data, the data might be considered MAR, where the missingness could be explained by observed factors, like country or year-specific reporting gaps.

Overall Interpretation:

Based on the results, both Country and Year variables seem to exhibit patterns of Missing at Random (MAR), rather than being Missing Completely at Random (MCAR). The 100% missing data for certain countries and years suggests that there may be systematic reasons for the missingness, which could be linked to issues like data reporting, country-specific factors, or missing data for particular years.

Thus, you can conclude that for both the Country and Year variables, the missingness is likely MAR and not completely random. However, the 100% missing data for certain years and countries might need more specific analysis or handling, such as removing or imputing those values, depending on their relevance to your analysis.







